hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapProxy.java

```diff
@@ -112,7 +112,7 @@ public void initialize() {
         if (!config.isAsyncFillup()) {
             for (int i = 0; i < nodeEngine.getPartitionService().getPartitionCount(); i++) {
                 ReplicatedRecordStore store = service.getReplicatedRecordStore(name, false, i);
-                while (!store.isLoaded()) {
+                while (store == null || !store.isLoaded()) {
```

Comment:
when we have a proxy store cannot be null i think

Reply:
The NPEs were thrown in this proxy. I still didn't figure out why it's not initialized by just changing the split-brain logic.

My theory is that we have a destroy() vs. clear() somewhere and on the lite member promotion we miss to call `ManagedService.init()`, which creates the stores for the ReplicatedMap.

Reply:
lite member cannot have replicated map, afaik

Reply:
Sorry, I meant `ReplicatedMapService.init()`. There is a skip for lite members:
```java
    @Override
    public void init(NodeEngine nodeEngine, Properties properties) {
        if (config.isLiteMember()) {
            return;
        }

        for (int i = 0; i < nodeEngine.getPartitionService().getPartitionCount(); i++) {
            partitionContainers[i] = new PartitionContainer(this, i);
        }
        nodeEngine.getExecutionService().getGlobalTaskScheduler().scheduleWithRepetition(new Runnable() {
            @Override
            public void run() {
                triggerAntiEntropy();
            }
        }, 0, SYNC_INTERVAL_SECONDS, TimeUnit.SECONDS);
    }
```
So I guess we clear those `partitionContainers` during the split-brain healing. I have to run the split-brain test with some break points :D

Reply:
Okay, what happens is that the `iterator.remove()` removes the collected `ReplicatedRecordStore` from the `PartitionContainer` on the merging node:
```
onPrepareMergeRunnableStart(): total stores in partition containers: 542
onPrepareMergeRunnableEnd(): total stores in partition containers: 0
```
This is fine, since the data has to be gone.

Then the `ReplicatedMapSplitBrainTest` tries to verify the values from the `ReplicatedMap` via `get()` calls on all proxies (one from the large cluster, one from the small cluster).
```java
    private void afterMergeDiscardMergePolicy() {
        assertEquals("value1", replicatedMapA1.get(key1));
        assertEquals("value1", replicatedMapA2.get(key1));
    }
```
This threw the NPE on the second call, which is the merged node from the smaller cluster.

The `get()` method calls `ReplicatedRecordStore store = service.getReplicatedRecordStore(getName(), false, partitionId)`. `false` is the parameter `create`:
```java
        PartitionContainer partitionContainer = partitionContainers[partitionId];
        if (create) {
            return partitionContainer.getOrCreateRecordStore(name);
        }
        return partitionContainer.getRecordStore(name);
```
So it is expected to return `null` here. The store is just recreated once the replication has been done again.

Reply:
I think the `ReplicatedMap` seems to be in general broken when a node is promoted from a lite member to a data member:
```java
@Test
public void testPromotedMember() {
    TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(2);
    HazelcastInstance instance1 = nodeFactory.newHazelcastInstance();
    ReplicatedMap<Integer, Integer> map1 = instance1.getReplicatedMap(randomName());
    map1.put(1, 1);

    Config config = new Config()
        .setLiteMember(true);

    HazelcastInstance instance2 = nodeFactory.newHazelcastInstance(config);
    instance2.getCluster().promoteLocalLiteMember();

    ReplicatedMap<Integer, Integer> map2 = instance2.getReplicatedMap(randomName());
    map2.get(1);
}
```
Results in:
```
com.hazelcast.replicatedmap.ReplicatedMapCantBeCreatedOnLiteMemberException: Can't create replicated map instance on [127.0.0.1]:5002
at com.hazelcast.replicatedmap.impl.ReplicatedMapService.createDistributedObject(ReplicatedMapService.java:246)
```
Okay, there are `config.isLiteMember()` checks everywhere, which should probably check the actual state of the member. After replacing that with `nodeEngine.getLocalMember().isLiteMember()` I run into the next issue:
```java
java.lang.NullPointerException
	at com.hazelcast.replicatedmap.impl.ReplicatedMapService.getReplicatedRecordStore(ReplicatedMapService.java:293)
```
This is since the `partitionContainers` are not initialized on the promotion.

Reply:
I'm working on a fix for this, but I guess I have to pull it out of this PR, since it's an existing problem with the `ReplicatedMap` and the lite member promotion, which also exists without split-brain scenarios.

Reply:
Just as a conclusion from the previous fix PR: The record store can be `null`, since it's removed by the `iterator.removed()` (and later destroyed) during the split-brain healing. So the `null` checks make sense. The stores are recreated as soon as the `ReplicatedMap` is synced again, but that happens with a delay. So the first `get()` calls from the split-brain test assertions ran into the NPE.